File: frontend/src/api/growth.ts
import { apiClient } from "./client";

export type GrowthStatus = "IMPROVING" | "STABLE" | "REQUIRING_ATTENTION";

export interface ConceptGrowthItem {
  conceptId: string;
  conceptName: string;
  status: GrowthStatus;
  masteryScore: number;
  confidence: number;
  trend: "IMPROVING" | "STABLE" | "DECLINING";
  changeScore: number | null;
  summary: string;
}

export interface ProjectGrowth {
  projectId: string;
  status: GrowthStatus;
  hasEvidence: boolean;
  overallMastery: number;
  averageConfidence: number;
  conceptsImproving: number;
  conceptsStable: number;
  conceptsRequiringAttention: number;
  assessedConcepts: number;
  assessmentCount: number;
  questionsAnswered: number;
  updatedAt: string | null;
  concepts: ConceptGrowthItem[];
}

export interface GrowthHistoryPoint {
  date: string;
  score: number;
  assessmentId: string;
}

interface BackendGrowth {
  project_id: string;
  status: GrowthStatus;
  has_evidence: boolean;
  overall_mastery: number;
  average_confidence: number;
  concepts_improving: number;
  concepts_stable: number;
  concepts_requiring_attention: number;
  assessed_concepts: number;
  assessment_count: number;
  questions_answered: number;
  updated_at: string | null;
  concepts: {
    concept_id: string;
    concept_name: string;
    status: GrowthStatus;
    mastery_score: number;
    confidence: number;
    trend: ConceptGrowthItem["trend"];
    change_score: number | null;
    summary: string;
  }[];
}

function toGrowth(g: BackendGrowth): ProjectGrowth {
  return {
    projectId: g.project_id,
    status: g.status,
    hasEvidence: g.has_evidence,
    overallMastery: g.overall_mastery,
    averageConfidence: g.average_confidence,
    conceptsImproving: g.concepts_improving,
    conceptsStable: g.concepts_stable,
    conceptsRequiringAttention: g.concepts_requiring_attention,
    assessedConcepts: g.assessed_concepts,
    assessmentCount: g.assessment_count,
    questionsAnswered: g.questions_answered,
    updatedAt: g.updated_at,
    concepts: (g.concepts ?? []).map((c) => ({
      conceptId: c.concept_id,
      conceptName: c.concept_name,
      status: c.status,
      masteryScore: c.mastery_score,
      confidence: c.confidence,
      trend: c.trend,
      changeScore: c.change_score,
      summary: c.summary,
    })),
  };
}

export const growthApi = {
  async get(projectId: string): Promise<ProjectGrowth> {
    const res = await apiClient.get<BackendGrowth>(`/api/v1/projects/${projectId}/growth`);
    return toGrowth(res.data);
  },

  async history(projectId: string): Promise<GrowthHistoryPoint[]> {
    const res = await apiClient.get<{
      items: { date: string; score: number; assessment_id: string }[];
    }>(`/api/v1/projects/${projectId}/growth/history`);
    return res.data.items.map((p) => ({
      date: p.date,
      score: p.score,
      assessmentId: p.assessment_id,
    }));
  },
};

File: backend/app/api/routes/growth.py
"""Growth endpoints: project aggregate, per-concept distribution, history.

The aggregate is computed live from persisted Mastery rows on every read
(never stale, nothing extra to migrate); per-concept Growth rows are the
persisted trail refreshed on assessment completion. Cold-start projects
return STABLE with has_evidence=False — never failure.
"""

from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.auth.dependencies import CurrentUser, get_owned_project
from app.db.session import get_db
from app.models.learning import Project
from app.schemas.common import Page
from app.schemas.growth import ConceptGrowthRead, GrowthHistoryPointRead, GrowthRead
from app.services.growth_service import GrowthService

router = APIRouter(prefix="/projects", tags=["growth"])

SessionDep = Annotated[Session, Depends(get_db)]
OwnedProject = Annotated[Project, Depends(get_owned_project)]


@router.get("/{project_id}/growth", response_model=GrowthRead)
def get_growth(project: OwnedProject, user: CurrentUser, session: SessionDep):
    service = GrowthService(session)
    aggregate = service.project_growth(user_id=user.id, project_id=project.id)
    concepts = service.concept_growth(user_id=user.id, project_id=project.id)
    return GrowthRead(
        project_id=project.id,
        status=aggregate.status,
        has_evidence=aggregate.has_evidence,
        overall_mastery=aggregate.overall_mastery,
        average_confidence=aggregate.average_confidence,
        concepts_improving=aggregate.concepts_improving,
        concepts_stable=aggregate.concepts_stable,
        concepts_requiring_attention=aggregate.concepts_requiring_attention,
        assessed_concepts=aggregate.assessed_concepts,
        assessment_count=aggregate.assessment_count,
        questions_answered=aggregate.questions_answered,
        updated_at=aggregate.updated_at,
        concepts=[
            ConceptGrowthRead(
                concept_id=c.concept_id,
                concept_name=c.concept_name,
                status=c.status,
                mastery_score=c.mastery_score,
                confidence=c.confidence,
                trend=c.trend,
                change_score=c.change_score,
                summary=c.summary,
            )
            for c in concepts
        ],
    )


@router.get("/{project_id}/growth/history", response_model=Page[GrowthHistoryPointRead])
def get_growth_history(
    project: OwnedProject,
    user: CurrentUser,
    session: SessionDep,
    limit: Annotated[int, Query(ge=1, le=200)] = 100,
):
    service = GrowthService(session)
    points = service.history(user_id=user.id, project_id=project.id, limit=limit)
    return Page[GrowthHistoryPointRead](
        items=[
            GrowthHistoryPointRead(date=p.date, score=p.score, assessment_id=p.assessment_id)
            for p in points
        ],
        total=len(points),
        page=1,
        page_size=limit,
    )
